src/cdk/table/table.ts

```diff
@@ -292,25 +300,45 @@ export class CdkTable<T> implements CollectionViewer {
     this._changeDetectorRef.markForCheck();
   }
 
-  /** Check for changes made in the data and render each change (row added/removed/moved). */
+  /**
+   * Check for changes made in the data and render each change (row added/removed/moved) and update
+   * row contexts.
+   */
   private _renderRowChanges() {
     const changes = this._dataDiffer.diff(this._data);
     if (!changes) { return; }
 
+    const insertOrMovedRowTuples: RowRecordViewTuple<T>[] = [];
+
     const viewContainer = this._rowPlaceholder.viewContainer;
     changes.forEachOperation(
-        (item: IterableChangeRecord<any>, adjustedPreviousIndex: number, currentIndex: number) => {
-          if (item.previousIndex == null) {
-            this._insertRow(this._data[currentIndex], currentIndex);
+        (record: IterableChangeRecord<T>, adjustedPreviousIndex: number, currentIndex: number) => {
+          if (record.previousIndex == null) {
+            const view = this._insertRow(record.item, currentIndex);
+            insertOrMovedRowTuples.push(new RowRecordViewTuple<T>(record, view));
           } else if (currentIndex == null) {
             viewContainer.remove(adjustedPreviousIndex);
           } else {
-            const view = viewContainer.get(adjustedPreviousIndex);
+            const view = <RowViewRef<T>>viewContainer.get(adjustedPreviousIndex);
             viewContainer.move(view!, currentIndex);
+            insertOrMovedRowTuples.push(new RowRecordViewTuple<T>(record, view));
           }
         });
 
-    this._updateRowContext();
+    // Update the row's implicit context for each inserted or moved row.
```

Comment:
"Update the $implicit property for the binding context for inserted and moved rows"